core/src/main/java/smile/regression/treeshap/ShapTree.java

```diff
@@ -0,0 +1,164 @@
+import smile.regression.RegressionTree;
+
+/** 
+ * structure for shap value calculation. implementation reference:
+ * 
+ * <pre>
+ * https://github.com/slundberg/shap/blob/master/shap/explainers/pytree.py
+ * </pre>
+ * 
+ * @author ray
+ */
+public class ShapTree {
+	
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ShapTree.class);
+
+  public boolean normalize = false;
+  public int[] children_left;
+  public int[] children_right;
+  public int[] children_default;
+  public int[] features;
+  public Node[] thresholds;
+  public double[] values;
+  public double[] node_sample_weight; // equals number of samples for a node divided by total number of samples
+  public int max_depth;
+
+  public ShapTree(RegressionTree tree) {
+    this(tree, false);
+  }
+
+  public ShapTree(RegressionTree tree, boolean normalize) {
+    this.normalize = normalize;
+
+    int totalSamples = tree.root().size();
+    children_left = new int[totalSamples];
+    children_right = new int[totalSamples];
+    features = new int[totalSamples];
+    thresholds = new Node[totalSamples];
+    values = new double[totalSamples];
+    node_sample_weight = new double[totalSamples];
+    //
+    // we convert smile tree structure to sth favored by shap calculation
+    //
+    Queue<Node> q = new ConcurrentLinkedQueue<Node>();
+    int idx = 0;
+    int nodeNum = 0;
+    q.add(tree.root());
+    Node n = q.poll();
+
+    while (n != null) {
+      node_sample_weight[idx] = (double) n.size() / (double) totalSamples;
+      
+      values[idx] = subtreeValues(n);
+      
+      thresholds[idx] = n;
+
+      if (n instanceof InternalNode) {
+    	  features[idx] = ((InternalNode) n).feature();
+          
+          q.add(((InternalNode) n).trueChild());
+          children_left[idx] = (++nodeNum);
+          
+          q.add(((InternalNode) n).falseChild());
+          children_right[idx] = (++nodeNum);
+      } else if (n instanceof LeafNode) {
+    	  features[idx] = -1;
+          children_left[idx] = -1;
+          children_right[idx] = -1;
+      } else {
+          String err = "error node type for tree structure conversion: " + n.getClass() + ":" + n.toString();
+          logger.error(err);
+          throw new RuntimeException(err);
+      }
+
+      n = q.poll();
+      idx++;
+    }
+
+    this.children_default = this.children_left; // missing values...
+
+    if (this.normalize) {
+        double valueSum = Arrays.stream(values).sum();
+        this.values = Arrays.stream(values).map(num -> num / valueSum).toArray();
+    } 
+
+    // we recompute the expectations to make sure they follow the SHAP logic
+    this.max_depth = computeExpectations(this.children_left, this.children_right, this.node_sample_weight, this.values, 0, 0);
+    // tracking progress for shap tree construction 
+    System.out.print(".");
+  }
+
+  /** calculate the constant prediction value of each node */
+  private double subtreeValues(Node n) {
```

Comment:
What do we need if it is a decision tree?

Reply:
As far as I understand, shap algorithm need the output from leaf node so that is why I only add RegressionTree in constructor. Not sure what we gonna do with DecisionTree since the leaf node could only generate a class label instead of some real-value output.

you may find more details at the original paper, page 3 to page 4
https://arxiv.org/pdf/1802.03888.pdf